File: java/src/main/java/com/example/cote/PGMRS/순위.java
package com.example.cote.PGMRS;

import java.io.*;
import java.nio.file.*;
import java.util.*;

public class 순위 {

  private static int countForward(int cur, boolean[][] graph, boolean[] visited) {
    int count = 1;
    visited[cur] = true;

    for (int i = 0; i < graph.length; i++) {
      if (!graph[cur][i] || visited[i]) continue;
      count += countForward(i, graph, visited);
    }

    return count;
  }

  private static int countBackward(int cur, boolean[][] graph, boolean[] visited) {
    int count = 1;
    visited[cur] = true;

    for (int i = 0; i < graph.length; i++) {
      if (!graph[i][cur] || visited[i]) continue;
      count += countBackward(i, graph, visited);
    }

    return count;
  }

  private static int solve(int n, int[][] edges) {

    boolean[][] graph = new boolean[n][n];

    for (int[] edge : edges) {
      graph[edge[0] - 1][edge[1] - 1] = true;
    }

    int count = 0;

    for (int i = 0; i < n; i++) {

      int wins = countForward(i, graph, new boolean[n]) - 1;
      int loses = countBackward(i, graph, new boolean[n]) - 1;

      if (wins + loses == n - 1) {
        count++;
      }
    }

    return count;
  }

  public static void main(String... args) throws IOException {

    Path p = Paths.get(System.getProperty("user.dir") + "/data/순위.txt");
    BufferedReader rd = Files.newBufferedReader(p);

    int testCases = Integer.parseInt(rd.readLine());

    while (testCases > 0) {

      int n = Integer.parseInt(rd.readLine());
      int numEdges = Integer.parseInt(rd.readLine());
      int[][] edges = new int[numEdges][2];

      for (int i = 0; i < numEdges; i++) {
        edges[i] = Arrays.stream(rd.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
      }

      int answer = Integer.parseInt(rd.readLine());

      System.out.println(solve(n, edges) + " " + answer);

      testCases--;
    }
  }
}

File: go/cote/PGMRS/순위.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func countBackard(cur int, graph [][]bool, visited []bool) int {
	ret := 1
	visited[cur] = true

	for i := 0; i < len(visited); i++ {
		if !visited[i] && graph[i][cur] {
			ret += countBackard(i, graph, visited)
		}
	}

	return ret
}

func countForward(cur int, graph [][]bool, visited []bool) int {
	ret := 1
	visited[cur] = true

	for i := 0; i < len(visited); i++ {
		if !visited[i] && graph[cur][i] {
			ret += countForward(i, graph, visited)
		}
	}

	return ret
}

func solve(n int, edges [][]int) int {
	graph := make([][]bool, n)

	for i := 0; i < n; i++ {
		graph[i] = make([]bool, n)
	}

	for _, edge := range edges {
		i, j := edge[0], edge[1]
		graph[i-1][j-1] = true
	}

	count := 0

	for i := 0; i < n; i++ {
		wins := countForward(i, graph, make([]bool, n)) - 1
		loses := countBackard(i, graph, make([]bool, n)) - 1

		if wins+loses == n-1 {
			count++
		}
	}

	return count
}

func main() {

	f, _ := os.Open("data/순위.txt")
	r := bufio.NewReader(f)

	s, _ := r.ReadString('\n')
	testcases, _ := strconv.Atoi(strings.Trim(s, "\n"))

	for it := 0; it < testcases; it++ {
		s, _ = r.ReadString('\n')
		n, _ := strconv.Atoi(strings.Trim(s, "\n"))

		s, _ = r.ReadString('\n')
		numEdges, _ := strconv.Atoi(strings.Trim(s, "\n"))
		edges := make([][]int, numEdges)

		for i := 0; i < numEdges; i++ {
			s, _ = r.ReadString('\n')
			edge := strings.Split(strings.Trim(s, "\n"), " ")
			edges[i] = make([]int, 2)

			for j := 0; j < 2; j++ {
				v, _ := strconv.Atoi(edge[j])
				edges[i][j] = v
			}
		}
		s, _ = r.ReadString('\n')
		answer, _ := strconv.Atoi(strings.Trim(s, "\n"))

		fmt.Printf("%d %d\n", answer, solve(n, edges))
	}
}
